pkg/settings/limits/rate_test.go

```diff
@@ -0,0 +1,272 @@
+	} else {
+		fmt.Println("5: success")
+	}
+
+	rl = GlobalRateLimiter(rate.Every(time.Second), 4) // reset
+
+	// Try 4
+	g = errgroup.Group{}
+	for range 4 {
+		g.Go(func() error {
+			return rl.AllowErr(ctx)
+		})
+	}
+	if err := g.Wait(); err != nil {
+		fmt.Println("4:", err)
+	} else {
+		fmt.Println("4: success")
+	}
+
+	// Output:
+	// 5: rate limited
+	// 4: success
+}
+
+func ExampleRateLimiter_Wait() {
+	ctx := context.Background()
+
+	rl := GlobalRateLimiter(rate.Every(time.Second), 1)
+	var wg sync.WaitGroup
+	wg.Add(2)
+	for range 2 {
+		go func() {
+			defer wg.Done()
+			start := time.Now()
+			err := rl.Wait(ctx)
+			if err != nil {
+				fmt.Println("rate limited:", err)
+				return
+			}
+			fmt.Println("waited:", time.Since(start).Round(time.Second))
+		}()
+	}
+	wg.Wait()
+
+	// Output:
+	// waited: 0s
+	// waited: 1s
+}
+
+func ExampleMultiRateLimiter() {
+	ctx := context.Background()
+	ctxA := contexts.WithCRE(ctx, contexts.CRE{Owner: "0xabcd"})
+	ctxB := contexts.WithCRE(ctx, contexts.CRE{Workflow: "ABCD"})
+
+	global := GlobalRateLimiter(rate.Every(time.Second), 4)
+	multiA := MultiRateLimiter{global, OwnerRateLimiter(rate.Every(time.Second), 4)}
+	multiB := MultiRateLimiter{global, WorkflowRateLimiter(rate.Every(time.Second), 4)}
+
+	// Try burst limit of 4 from A
+	var g errgroup.Group
+	for range 4 {
+		g.Go(func() error {
+			return multiA.AllowErr(ctxA)
+		})
+	}
+	if err := g.Wait(); err != nil {
+		fmt.Println("A:", err)
+	} else {
+		fmt.Println("A: success")
+	}
+
+	// Try burst limit of 4 from A & B at the same time
+
+	g = errgroup.Group{}
+	for range 4 {
+		g.Go(func() error {
+			return multiA.AllowErr(ctxA)
+		})
+	}
+	for range 4 {
+		g.Go(func() error {
+			return multiB.AllowErr(ctxB)
+		})
+	}
+	if err := g.Wait(); err != nil {
+		fmt.Println("A&B:", err)
+	} else {
+		fmt.Println("A&B: success")
+	}
+
+	// Output:
+	// A: success
+	// A&B: rate limited
+}
+
+func TestFactory_NewRateLimiter(t *testing.T) {
+	t.Parallel()
+	for _, tt := range []struct {
+		scope settings.Scope
+		cre   contexts.CRE
+	}{
+		{settings.ScopeGlobal, contexts.CRE{}},
+		{settings.ScopeWorkflow, contexts.CRE{Workflow: "wf-id"}},
+	} {
+		t.Run(tt.scope.String(), func(t *testing.T) {
+			t.Parallel()
+			mc := newMetricsChecker(t)
+			f := Factory{Meter: mc.Meter(t.Name())}
+			s := settings.Rate(rate.Every(30*time.Second), 100)
+			s.Key = "foo.bar"
+			s.Scope = tt.scope
+			rl, err := f.NewRateLimiter(s)
+			require.NoError(t, err)
+			t.Cleanup(func() { assert.NoError(t, rl.Close()) })
+
+			ctx := t.Context()
+			ctx = contexts.WithCRE(ctx, tt.cre)
+
+			assert.NoError(t, rl.AllowErr(ctx))
+
+			assert.Error(t, rl.AllowNErr(ctx, time.Now(), 101))
+
+			time.Sleep(2 * pollPeriod) // ensure at least one update
+
+			attrs := attribute.NewSet(kvsFromScope(ctx, tt.scope)...)
+			ms := mc.lastResourceFirstScopeMetric(t)
+			require.Equal(t, metrics{
+				{
+					Name: "rate.foo.bar.limit",
+					Data: metricdata.Gauge[float64]{
```

Comment:
I will fill in unit and description too:
```suggestion
					Name: "rate.foo.bar.limit",
					Unit: "",
					Description "",
					Data: metricdata.Gauge[float64]{
```